### ISSUE TITLE
Create an HTML/text export generator

### ISSUE THREAD

**Issue by spencerwahl** [fgpv-vpgf/fgpv-vpgf#3598]
Create an HTML to SVG/canvas generator for export. https://html2canvas.hertzen.com/ should be tried out. This will allow easy generation of fancy text blocks for map export.

If this cannot be done, convert the footnote generator into something more generic.

**Comment by AleksueiR** [comment]
Extremely slow progress due to the ongoing issues with my laptop - a RAMP build takes 10+ minutes to complete.

🐌 🐌 💻 🐌 🐌

**Comment by AleksueiR** [comment]
## Export Anatomy

The export image comprises a number of vertically-stacked pieces of canvas returned by the `Export Component`s. Each `Export Component` can use one of more `Export Generator`s to create images which later will be fused together. The canvas returned by an `Export Component` must have the width of the viewer container (or simply the width of the map).

For example, here's the map `Export Component` configuration:

```js
map: {
    generators: [
        exportGenerators.mapDummyGenerator,
        exportGenerators.mapSVGGenerator,
        exportGenerators.mapImageGenerator
    ],
    graphicOrder: [0, 2, 1]
}
```

### Current Behaviour

In the image below:

- grey boxes contain actual export content (title, map image, northarrow, etc.)
- black-bordered boxes around grey ones represent `Export Generator`s output
  - notice that they have the full width of the export image - they are not sized to simply fit their content
  - where several outputs are stacked, they will be fused together by the `Export Component` which called these generators
- dotted lines separate outputs of the `Export Component`s from each other
  - the size of the canvas returned by `Export Generator`s using by an `Export Components` must be exactly the same because they will be overlaid on top of each other in the order specified by the `graphicOrder` and fused

![](https://i.imgur.com/ULVkFuP.png)

Take the `mapElements` component for example. It uses two generators to create images of the scalebar and northarrow. These generators create strips of canvas and position the corresponding images on the sides of the strips. 

![](https://i.imgur.com/EHEUbOb.png)

The parent `Export Component,` merges them to create its output which will be used in the final image.

![](https://i.imgur.com/fFpSmhm.png)

With export plugins getting direct access to the export generators, it becomes necessary to rejig how export components and generators work. Using the current approach, a plugin calling a northarrow generator will get a strip of canvas which does not correspond directly to the size of the arrow itself. Since the plugin won't be able to tell the size of the arrow, it won't be able to custom-position the arrow relative to the export canvas. Hence the refactor.

### New Behaviour

In the image below:

- grey, black-bordered boxes now represent the `Export Generator` output (title, map image, northarrow, etc.)
  - notice they are sized to fit the content now
- boxes with dashed black borders represent `Export Component` canvas output
  - notice that they are still sized to the full width of the export image
  - individual generator outputs will be fused to the `Export Component`'s strip of canvas
- dotted lines separate outputs of the `Export Component`s from each other![]

![](https://i.imgur.com/FDWAhXZ.png)

`Export Component`s will have an additional configuration option - `graphicPosition`. For example:

```js
mapElements: {
    generators: [exportGenerators.scalebarGenerator, exportGenerators.northarrowGenerator],
    // position scalebar on the left, and north-arrow on the right side of the export image
    graphicPosition: [
        {
            justify: 'start',
            align: 'center'
        },
        {
            justify: 'end',
            align: 'center'
        }
    ]
}
```

`Export Component`s always return a canvas piece whose width corresponds to the width of the viewer container and whose height fits the tallest output of the generator used by this component. When an `Export Compoent` receives output of its `Export Generator`s, it positions them according to the `graphicPosition` option.

This way, any export plugin calling any of the internal export generators, will get a canvas sized to its content, and can easily position it relative to the canvas of the resulting export image.

**Note**: this new behaviour only affect the internal workings of the export service, and does not require any changes to the config, schema, or change the appearance of the internally generated export image. 

**tldr**: now `Export Componet`s are responsible for positioning outputs of `Export Generator`s.
